### ISSUE TITLE
Inability to process x86 assemblies

### ISSUE THREAD

**Issue by akamyshanov** [RSuter/NSwag#873]
Neither NSwag Studio nor command line utility can open x86 assemblies. Is this by design or can be somehow fixed?

Command-line utility stacktrace:
```
PS C:\Workspace\Projects\***> ..\packages\NSwag.MSBuild.11.3.3\build\NSwag.exe run nswag.json
NSwag command line tool for .NET 4.6+, toolchain v11.3.3.0 (NJsonSchema v9.4.2.0) (x64)
Visit http://NSwag.org for more information.
NSwag bin directory: C:\Workspace\Projects\***\packages\NSwag.MSBuild.11.3.3\build

Executing file 'nswag.json'...
System.BadImageFormatException: Could not load file or assembly 'file:///C:\Workspace\Projects\***\***\bin\Release\***.exe' or one of its dependencies. An attempt was made to load a program with an incorrect format.
File name: 'file:///C:\Workspace\Projects\***\***\bin\Release\***.exe'

Server stack trace:
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.LoadFrom(String assemblyFile)
   at NSwag.SwaggerGeneration.WebApi.WebApiAssemblyToSwaggerGenerator.WebApiAssemblyLoader.<>c.<GetControllerTypesAsync>b__2_0(String path)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at NSwag.SwaggerGeneration.WebApi.WebApiAssemblyToSwaggerGenerator.WebApiAssemblyLoader.<GetControllerTypesAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.SwaggerGeneration.WebApi.WebApiAssemblyToSwaggerGenerator.WebApiAssemblyLoader.<GenerateForControllersAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.SwaggerGeneration.WebApi.WebApiAssemblyToSwaggerGenerator.WebApiAssemblyLoader.GenerateForControllers(IEnumerable`1 controllerClassNames, String settingsData)
   at System.Runtime.Remoting.Messaging.StackBuilderSink._PrivateProcessMessage(IntPtr md, Object[] args, Object server, Object[]& outArgs)
   at System.Runtime.Remoting.Messaging.StackBuilderSink.SyncProcessMessage(IMessage msg)

Exception rethrown at [0]:
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at NSwag.SwaggerGeneration.WebApi.WebApiAssemblyToSwaggerGenerator.WebApiAssemblyLoader.GenerateForControllers(IEnumerable`1 controllerClassNames, String settingsData)
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.SwaggerGeneration.WebApi.WebApiAssemblyToSwaggerGenerator.<GenerateForControllersAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.WebApiToSwaggerCommandBase.<<RunAsync>b__77_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.WebApiToSwaggerCommandBase.<RunAsync>d__77.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.WebApiToSwaggerCommandBase.<RunAsync>d__76.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.NSwagDocumentBase.<GenerateDocumentAsync>d__29.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.NSwagDocumentBase.<ExecuteAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.Document.ExecuteDocumentCommandBase.<ExecuteDocumentAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.Document.ExecuteDocumentCommandBase.<RunAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NConsole.CommandLineProcessor.<ProcessSingleAsync>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NConsole.CommandLineProcessor.<ProcessAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NConsole.CommandLineProcessor.Process(String[] args, Object input)
   at NSwag.Commands.NSwagCommandProcessor.Process(String[] args)

WRN: Assembly binding logging is turned OFF.
To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.
Note: There is some performance penalty associated with assembly bind failure logging.
To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].
```

NSwag Studio screenshot:
<img width="986" alt="2017-08-09 at 15 15" src="https://user-images.githubusercontent.com/1358330/29121355-e4b80162-7d16-11e7-9e34-9072d9644d13.png">

**Comment by RSuter** [comment]
It seems that a x64 process (NSwag by default) cannot load x86 DLLs - only Any CPU and x64. If you know a way to solve this problem please share it here? Or is it possible to start a .NET process as x86?

**Comment by akamyshanov** [comment]
Correct. Unfortunately I don't think there's a clean and easy way except creating separate binaries targeting x86. E.g. `NSwag.x86.exe`. Do you think that's possible?
As an option, if you are against creating separate build configurations for x86, there's a [CorFlags](https://docs.microsoft.com/en-us/dotnet/framework/tools/corflags-exe-corflags-conversion-tool) utility that modifies built AnyCPU executables to run as x86. So, it can be integrated into the build process of NSwag.

**Comment by RSuter** [comment]
NSwagStudio and NSwagConsole are already built with AnyCPU but with "Prefer 32 bit" disabled.

**Comment by stijnherreman** [comment]
You'll need to create separate binaries to be able to load x86-only or x64-only binaries. See https://stackoverflow.com/a/5229805/247702

@akamyshanov I assume you have your reasons for targetting x86 instead of AnyCPU?

**Comment by RSuter** [comment]
How about using something like that in the NPM package if a specific flag is set?

http://coolsoft.altervista.org/en/blog/2012/01/run-anycpu-net-applications-x86-mode

**Comment by akamyshanov** [comment]
@stijnherreman Exactly :) x86-only dependencies using 32-bit native DLLs.

My current workaround is to build AnyCPU (and ignore the produced warnings) just to feed it to NSwag.

@RSuter Also a valid workaround, but having a separate x86 NSwag binary seems cleaner to me. Maybe I'm not the only one with such a problem.

**Comment by RSuter** [comment]
And how can you choose which version of the .exe to run if both binaries are available?

- MSBuild => Choose between $(NSwagExe) and $(NSwagExe_x86)
- NPM => Add `--x86` option
- NSwagStudio => ??

**Comment by RSuter** [comment]
What do you think: 

https://github.com/RSuter/NSwag/pull/876

?

**Comment by akamyshanov** [comment]
@RSuter Wow, that was fast! 👍 
Cannot say anything about the NPM part but the console and MSBuild parts seem ok to me.
BTW, why create a new project instead of adding a new x86 configuration to the existing one? In order not to build the solution twice?

**Comment by RSuter** [comment]
How can we do that? I thought only multiple target frameworks are supported?

**Comment by akamyshanov** [comment]
@RSuter 
https://github.com/akamyshanov/NSwag/commit/9fc4e89be9a0b567a81159d9b399789cb4190469

However, you would need to build the solution twice in order to get both build results, I thought that's why you created a new project in the first place.

**Comment by RSuter** [comment]
Hmm, if i need to build it twice i think its better as is because it is much harder to integrate in the automated builds...

**Comment by akamyshanov** [comment]
I agree. Building the whole solution twice just for one project is not the best option from the automated build point of view. Even though it's less "code" to maintain.

Thank you for the quick reaction, will wait for the next release with these changes included.

**Comment by RSuter** [comment]
Released with v11.3.5
